apps/src/studio/studio.js

```diff
@@ -2896,23 +2911,25 @@ Studio.allGoalsVisited = function() {
 };
 
 var checkFinished = function () {
-  // if we have a succcess condition and have accomplished it, we're done and successful
-  if (level.goal && level.goal.successCondition && level.goal.successCondition()) {
+
+  var hasGoals = Studio.spriteGoals_.length !== 0;
+  var achievedGoals = Studio.allGoalsVisited();
+  var hasSuccessCondition = level.goal && level.goal.successCondition ? true : false;
+  var achievedOptionalSuccessCondition = !hasSuccessCondition || utils.valueOr(level.goal.successCondition(), true);
+  var achievedRequiredSuccessCondition = hasSuccessCondition && utils.valueOr(level.goal.successCondition(), false);
+
+  if ((hasGoals && achievedGoals && achievedOptionalSuccessCondition) ||
+      (hasGoals && level.completeOnSuccessConditionNotGoals && achievedRequiredSuccessCondition) ||
+      (!hasGoals && achievedRequiredSuccessCondition)) {
```

Comment:
Would love to see a comment somewhere on exactly what completeOnSuccessConditionNotGoals means. Not sure what the best place for that is. Maybe here?